### ISSUE TITLE
[FEA] Implement CPU-only version of NVTabular

### ISSUE THREAD

**Issue by rjzamora** [NVIDIA/NVTabular#534]
**Is your feature request related to a problem? Please describe.**
NVTabular currently requires that cudf be installed, and that the user has 1+ NVIDIA GPUs in their system.  This makes it difficult for users without direct/local access to GPU-equipped systems to iteratively develop and test new models/workflows.  It also makes it difficult to directly compare GPU and CPU performance.  

**Describe the solution you'd like**
All components of NVTabular should support "CPU Mode", which should be able to function without using `cudf` in any way.  An end-to-end solution likely involves the following features/improvements:

- [x] The `Dataset` API should handle a `cpu` boolean argument with a default value of `False` if `cudf` is installed (and `True` otherwise).  The underlying dataset engines should honor this option when the internal `ddf` is generated (using pandas-backed `dask.dataframe` if `cpu=True`, and `dask_cudf` otherwise). **[Covered in #556]**
- [x] `Dataset` should offer `to_gpu` and `to_cpu` methods to allow the user to explicitly move between host and device memory. **[Covered in #556]**
- [x] All operators should use a dispatching approach based on the location of the data that is passed to it.  That is, a pandas-backed partition should stay on the CPU, and a cudf-backed partition should stay on the GPU.  **This is likely to be the lions share of the work**, but it doesn't need to be done in one sweeping PR.  Once the `Dataset` work is done, the ops can be knocked out in parallel.
    - [x] Bucketize [#874]
    - [x] Categorify [#566]
    - [x] Categorify 
    - [x] Clip [#566]
    - [x] ColumnSimilarity [#880]
    - [x] ColumnSimilarity 
    - [x] DataStats [#884]
    - [x] DifferenceLag  [#884]
    - [x] Dropna [#884]
    - [x] FillMissing [#566] [#883]
    - [x] FillMedian [#566]
    - [x] Filter [#879]
    - [x] HashBucket [#884]
    - [x] HashedCross [#884]
    - [x] JoinExternal [#843]
    - [x] JoinGroupby [#877]
    - [x] LambdaOp [#883] [Mostly covered in #566]
    - [x] ListSlice 
    - [x] LogOp [#566]
    - [x] Normalize [#561]
    - [x] NormalizeMinMax [#883] [#561] 
    - [x] Rename [#883]
    - [x] TargetEncoding [#876]
- [x] `Dataset.to_parquet` should honor the `Dataset.cpu` attribute and avoid the use of cudf (i.e. CPU mem/compute) when writing out a processed Dataset.  (Not sure how important `to_hugectr` is?)
- [x] ~~Hashing should return same results between cpu/gpu (#569)~~
- [x] ~~Triton inference support should be able to select running on cpu/gpu (#739)
- [x] cudf should be an optional dependency for cpu-only systems. [#894 ]
- [x] We should add CI for cpu only support - including CI testing NVTabular on systems like windows/osx/linux (#738)
- [x] Dataloader support #856  (Julio)
  - [x] PyTorch cpu dataloader (#862)
  - [x] TensorFlow cpu dataloader

**Describe alternatives you've considered**
The alternative is to use pandas/dask-dataframe instead of NVTabular.

**Comment by rjzamora** [comment]
At this point, cpu support has already been added for many of the `Operator` definitions.  Here are the steps others can take to cover the remaining ops:


## Step 1: Remove all direct usage of `dask_cudf` from `ops/<operator-name>.py`

To make it easier to enable NVTabular to run without a cudf installation in the future, we avoid importing and/or using dask_cudf directly.  To do this, we leverage the fact that a `dask_cudf.DataFrame` is actually an instance of `dask.dataframe.DataFrame`.  Therefore, we can change something like…

```python
    def fit(self, columns: ColumnNames, ddf: dask_cudf.DataFrame):
```

To…

```python
    def fit(self, columns: ColumnNames, ddf: dd.DataFrame):
```

Where `dd` is a common import alias for `dask.dataframe`.

## Step 2: Remove all direct usage of `cudf` from `ops/<operator-name>.py`

To make it easier to enable NVTabular to run without a `dask_cudf` installation in the future, we avoid importing and/or using `cudf` directly.  To do this, we use the general `DataFrameType` for type-checking, and move any `cudf`-specific logic into the `nvtabular.dispatch` module.  Since the operator logic may be applied to either a cuff-backed DataFrame **or** and pandas-backed DataFrame, we also avoid using the variable name `gdf` (in favor of `df`).  Therefore, a typical `transform` definition will be changed from something like…

```python
    def transform(self, columns: ColumnNames, gdf: cudf.DataFrame) -> cudf.DataFrame:
        …
        return gdf
```

To…

```python
    def transform(self, columns: ColumnNames, df: DataFrameType) -> DataFrameType:
        …
        return df
```

## Step 3: Remove all direct usage of `cupy` from `ops/<operator-name>.py`

To make it easier to enable NVTabular to run without a `cupy` installation in the future, we avoid importing and/or using `cupy` directly.  To do this, we move any `cupy`-specific logic into the `nvtabular.dispatch` module.  We also leverage automatic [NEP-18](https://numpy.org/neps/nep-0018-array-function-protocol.html)-based dumpy dispatching.  

## Step 4: Add cpu=True test coverage

After removing all direct references to cudf/dask_cudf/cupy in the operator logic, you must add test coverage for cpu support. This step should be fairly straightforward, unless the existing test is making many backend-specific assumptions (in which case you may need to modify the logic of the existing test).  In general, however, the test will end up looking something like…

```python
@pytest.mark.parametrize("cpu", [True, False])
def test_<operator-name>(keys, cpu):
    …
    dataset = nvt.Dataset(<input-data>, cpu=cpu)
    …
    workflow.fit_transform(dataset)
    …
```

## General Tips

There is a good chance that you will need to leverage existing features in `dispatch`.  Some particularly useful utilities include `_arange`, `_arrray`, and `_concat_columns`.  The first two of these utilities can be used to generate an appropriate numpy/cupy array without explicitly “choosing” cupy or numpy. The dispatching mechanism requires you to pass in a reference DataFrame to the `like_df=` argument.  If you pass in a pandas DataFrame, vanilla lumpy will be used, otherwise, a cupy array will be generated.  The `_concat_columns` utility is a dispatch for `cudf/pd.concat(dfs, axis=1)`, for row-wise concatenation (`cudf/pd.concat(dfs, axis=0)`), use `dask.dataframe.core._concat`

cc @albert17 (in case this is useful)

**Comment by benfred** [comment]
CPU support has been added - including CI on cpu only systems ! https://github.com/NVIDIA/NVTabular/pull/927.
